app/src/components/carboneCopyApi.js

```diff
@@ -0,0 +1,119 @@
+const mime = require('mime-types');
+const path = require('path');
+const Problem = require('api-problem');
+const telejson = require('telejson');
+
+const carboneRender = require('./carboneRender');
+const FileCache = require('./fileCache');
+const { truthy } = require('./utils');
+
+const fileCache = new FileCache();
+
+const carboneCopyApi = {
+  init() {
+    carboneRender.carboneSet();
+  },
+
+  findAndRender: (hash, req, res) => {
+    const template = fileCache.find(hash);
+    if (!template.success) {
+      return new Problem(template.errorType, { detail: template.errorMsg }).send(res);
+    } else {
+      return carboneCopyApi.renderTemplate(template, req, res);
+    }
+  },
+
+  getFromCache: async (hash, hashHeaderName, download, remove, res) => {
+    const file = fileCache.find(hash);
+    if (!file.success) {
+      return new Problem(file.errorType, { detail: file.errorMsg }).send(res);
+    }
+
+    let cached = undefined;
+    if (download) {
+      try {
+        cached = await fileCache.read(hash);
+      } catch (e) {
+        return new Problem(500, { detail: e.message }).send(res);
+      }
+    }
+
+    if (remove) {
+      const removed = await fileCache.remove(hash);
+      if (!removed.success) {
+        return new Problem(removed.errorType, { detail: removed.errorMsg }).send(res);
+      }
+    }
+
+    res.setHeader(hashHeaderName, file.hash);
+    if (cached) {
+      res.setHeader('Content-Disposition', `attachment; filename=${file.name}`);
+      res.setHeader('Content-Transfer-Encoding', 'binary');
+      res.setHeader('Content-Type', mime.contentType(path.extname(file.name)));
+      res.setHeader('Content-Length', cached.length);
+      return res.send(cached);
+    } else {
+      return res.sendStatus(200);
+    }
+  },
+
+  renderTemplate: async (template, req, res) => {
+    let data = req.body.data;
+    let options = {};
+    let formatters = {};
+
+    try {
+      options = req.body.options;
+    } catch (e) {
+      return new Problem(400, { detail: 'options not provided or formatted incorrectly' }).send(res);
+    }
+
+    options.convertTo = options.convertTo || template.ext;
+    if (options.convertTo.startsWith('.')) {
+      options.convertTo = options.convertTo.slice(1);
+    }
+
+    options.reportName = options.reportName || `${path.parse(template.name).name}.${options.convertTo}`;
+    // ensure the reportName has the same extension as the convertTo...
+    if (options.convertTo !== path.extname(options.reportName).slice(1)) {
+      options.reportName = `${path.parse(options.reportName).name}.${options.convertTo}`;
+    }
+
+    if (typeof data !== 'object' || data === null) {
+      try {
+        data = req.body.data;
+      } catch (e) {
+        return new Problem(400, { detail: 'data not provided or formatted incorrectly' }).send(res);
+      }
+    }
+
+    try {
+      formatters = telejson.parse(req.body.formatters);
+      // eslint-disable-next-line no-empty
+    } catch (e) {
```

Comment:
should we add some exception action here and not keep it empty?

Reply:
In the event the formatters can't be parsed, it'll default to not do anything special with formatting. Even if we printed out some warning, there wouldn't be too much we'd be able to act on, as it'd mostly be caused by mis-formatted requests. I'll flag it as a TODO in the next PR so that when we get back to the JSON logging overhaul, we can revisit this.